***Задание 1. Базовое изучение***

Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [2]:
import pandas as pd
import numpy as np

Открываем датасет, изучаем описание его столбцов, называем столбцы согласно описанию

In [3]:
horses_data = pd.read_csv('horse_data.csv', header=None, na_values='?', 
                        names=['surgery',  'age',  'case_number', 'rectal_temperature', 'pulse', 'respiratory_rate', 
                               'temperature_of_extremities', 'peripheral_pulse', 'mucous_membranes', 'capillary_refill_time',
                               'pain', 'peristalsis', 'abdominal_distension', 'nasogastric_tube', 'nasogastric_reflux',
                               'nasogastric_reflux PH', 'rectal_examination', 'abdomen', 'packed_cell_volume', 'total_protein',
                               'abdominocentesis_appearance', 'abdomcentesis_total_protein', 'outcome', 'surgical_lesion',
                               'type_of_lesion_25', 'type_of_lesion_26', 'type_of_lesion_27', 'cp_data'])
horses_data.head()

surgery  age  case_number  rectal_temperature  pulse  respiratory_rate  \
0      2.0    1       530101                38.5   66.0              28.0   
1      1.0    1       534817                39.2   88.0              20.0   
2      2.0    1       530334                38.3   40.0              24.0   
3      1.0    9      5290409                39.1  164.0              84.0   
4      2.0    1       530255                37.3  104.0              35.0   

   temperature_of_extremities  peripheral_pulse  mucous_membranes  \
0                         3.0               3.0               NaN   
1                         NaN               NaN               4.0   
2                         1.0               1.0               3.0   
3                         4.0               1.0               6.0   
4                         NaN               NaN               6.0   

   capillary_refill_time  ...  packed_cell_volume  total_protein  \
0                    2.0  ...                45.0            8.4   
1                    1.0  ...                50.0           85.0   
2                    1.0  ...                33.0            6.7   
3                    2.0  ...                48.0            7.2   
4                    2.0  ...                74.0            7.4   

   abdominocentesis_appearance  abdomcentesis_total_protein  outcome  \
0                          NaN                          NaN      2.0   
1                          2.0                          2.0      3.0   
2                          NaN                          NaN      1.0   
3                          3.0                          5.3      2.0   
4                          NaN                          NaN      2.0   

   surgical_lesion  type_of_lesion_25  type_of_lesion_26  type_of_lesion_27  \
0                2              11300                  0                  0   
1                2               2208                  0                  0   
2                2                  0                  0                  0   
3                1               2208                  0                  0   
4                2               4300                  0                  0   

   cp_data  
0        2  
1        2  
2        1  
3        1  
4        2  

[5 rows x 28 columns]

In [25]:
horses_data.shape

(300, 28)

Выбираем столбцы для дальнейшего изучения, например: 'case_number', 'age', 'surgery', 'rectal_temperature', 'pulse', 'mucous_membranes', 'abdominal_distension', 'packed_cell_volume', 'total_protein', 'outcome'

In [4]:
horse_stat = horses_data.filter(['case_number', 'age', 'surgery', 'rectal_temperature', 'pulse', 'mucous_membranes', 
                                 'abdominal_distension', 'packed_cell_volume', 'total_protein', 'outcome'])
horse_stat.head()

case_number  age  surgery  rectal_temperature  pulse  mucous_membranes  \
0       530101    1      2.0                38.5   66.0               NaN   
1       534817    1      1.0                39.2   88.0               4.0   
2       530334    1      2.0                38.3   40.0               3.0   
3      5290409    9      1.0                39.1  164.0               6.0   
4       530255    1      2.0                37.3  104.0               6.0   

   abdominal_distension  packed_cell_volume  total_protein  outcome  
0                   4.0                45.0            8.4      2.0  
1                   2.0                50.0           85.0      3.0  
2                   1.0                33.0            6.7      1.0  
3                   4.0                48.0            7.2      2.0  
4                   NaN                74.0            7.4      2.0

* ***Анализ столбцов*** 

In [5]:
#Проверяем пропуски

for column in horse_stat:
    missings = horse_stat[column].isnull().mean()
    print(f'{column} - {missings :.1%}')

case_number - 0.0%
age - 0.0%
surgery - 0.3%
rectal_temperature - 20.0%
pulse - 8.0%
mucous_membranes - 15.7%
abdominal_distension - 18.7%
packed_cell_volume - 9.7%
total_protein - 11.0%
outcome - 0.3%


In [6]:
horse_stat_m = horse_stat.filter(['rectal_temperature', 'pulse', 'packed_cell_volume', 'total_protein'])
horse_stat_m.head()

rectal_temperature  pulse  packed_cell_volume  total_protein
0                38.5   66.0                45.0            8.4
1                39.2   88.0                50.0           85.0
2                38.3   40.0                33.0            6.7
3                39.1  164.0                48.0            7.2
4                37.3  104.0                74.0            7.4

In [7]:
#Базовые метрики числовых показателей

horse_stat_d = horse_stat_m.describe()

metrics = {}
for column in horse_stat_m:
    metrics[column] = horse_stat_m[column].mode()[0]
horse_stat_d = pd.concat([horse_stat_d, pd.DataFrame([metrics], index = ['mode'])])

metrics = {}
for column in horse_stat_m:
    metrics[column] = horse_stat_m[column].max() - horse_stat_m[column].min()
horse_stat_d = pd.concat([horse_stat_d, pd.DataFrame([metrics], index = ['range'])])

metrics = {}
for column in horse_stat_m:
    metrics[column] = horse_stat_m[column].std()
horse_stat_d = pd.concat([horse_stat_d, pd.DataFrame([metrics], index = ['STD'])])

metrics = {}
for column in horse_stat_m:
    metrics[column] = horse_stat_m[column].var()
horse_stat_d = pd.concat([horse_stat_d, pd.DataFrame([metrics], index = ['variation'])])

horse_stat_d

rectal_temperature       pulse  packed_cell_volume  total_protein
count              240.000000  276.000000          271.000000     267.000000
mean                38.167917   71.913043           46.295203      24.456929
std                  0.732289   28.630557           10.419335      27.475009
min                 35.400000   30.000000           23.000000       3.300000
25%                 37.800000   48.000000           38.000000       6.500000
50%                 38.200000   64.000000           45.000000       7.500000
75%                 38.500000   88.000000           52.000000      57.000000
max                 40.800000  184.000000           75.000000      89.000000
mode                38.000000   48.000000           37.000000       6.500000
range                5.400000  154.000000           52.000000      85.700000
STD                  0.732289   28.630557           10.419335      27.475009
variation            0.536247  819.708775          108.562532     754.876145

Заменим числовые значения столбцов 'age', 'surgery', 'mucous_membranes', 'abdominal_distension', 'outcome'. Данные получены из приложенного описания 

In [8]:
horse_stat.loc[horse_stat['age'] == 1, 'age'] = 'Adult'
horse_stat.loc[horse_stat['age'] == 9, 'age'] = 'Young'

horse_stat.loc[horse_stat['surgery'] == 1, 'surgery'] = 'Yes'
horse_stat.loc[horse_stat['surgery'] == 2, 'surgery'] = 'No'

horse_stat.loc[horse_stat['mucous_membranes'] == 1, 'mucous_membranes'] = 'normal pink'
horse_stat.loc[horse_stat['mucous_membranes'] == 2, 'mucous_membranes'] = 'bright pink'
horse_stat.loc[horse_stat['mucous_membranes'] == 3, 'mucous_membranes'] = 'pale pink'
horse_stat.loc[horse_stat['mucous_membranes'] == 4, 'mucous_membranes'] = 'pale cyanotic'
horse_stat.loc[horse_stat['mucous_membranes'] == 5, 'mucous_membranes'] = 'bright red / injected'
horse_stat.loc[horse_stat['mucous_membranes'] == 6, 'mucous_membranes'] = 'dark cyanotic'

horse_stat.loc[horse_stat['abdominal_distension'] == 1, 'abdominal_distension'] = 'none'
horse_stat.loc[horse_stat['abdominal_distension'] == 2, 'abdominal_distension'] = 'slight'
horse_stat.loc[horse_stat['abdominal_distension'] == 3, 'abdominal_distension'] = 'moderate'
horse_stat.loc[horse_stat['abdominal_distension'] == 4, 'abdominal_distension'] = 'severe'

horse_stat.loc[horse_stat['outcome'] == 1, 'outcome'] = 'lived'
horse_stat.loc[horse_stat['outcome'] == 2, 'outcome'] = 'died'
horse_stat.loc[horse_stat['outcome'] == 3, 'outcome'] = 'was euthanized'

horse_stat.head()

case_number    age surgery  rectal_temperature  pulse mucous_membranes  \
0       530101  Adult      No                38.5   66.0              NaN   
1       534817  Adult     Yes                39.2   88.0    pale cyanotic   
2       530334  Adult      No                38.3   40.0        pale pink   
3      5290409  Young     Yes                39.1  164.0    dark cyanotic   
4       530255  Adult      No                37.3  104.0    dark cyanotic   

  abdominal_distension  packed_cell_volume  total_protein         outcome  
0               severe                45.0            8.4            died  
1               slight                50.0           85.0  was euthanized  
2                 none                33.0            6.7           lived  
3               severe                48.0            7.2            died  
4                  NaN                74.0            7.4            died

In [9]:
#Для количественных показателей возможно получить моду
print('Мода age:', horse_stat.age.mode()[0])
print('Мода surgery:', horse_stat.surgery.mode()[0])
print('Мода mucous_membranes:', horse_stat.mucous_membranes.mode()[0])
print('Мода abdominal_distension:', horse_stat.abdominal_distension.mode()[0])
print('Мода outcome:', horse_stat.outcome.mode()[0])

Мода age: Adult
Мода surgery: Yes
Мода mucous_membranes: normal pink
Мода abdominal_distension: none
Мода outcome: lived


***Краткое описание данных***

* ***case_number*** - категориальный идентификатор лошади, может встречаться в данных несколько раз, если лошадь проходила повторное лечение (таких случаев около 6% из всех наблюдений). Пропуски в данных отсутствуют.

In [34]:
horse_stat['case_number'].value_counts().value_counts(normalize=True)

1    0.943662
2    0.056338
Name: case_number, dtype: float64

* ***age*** - категориальные значения. Согласно описанию категории, есть два варианта значения возраста - '1':"взрослая лошадь" или '2':"молодая < 6 месяцев", но '2' не встречается ни разу, но есть многократно повторяющееся значение '9' - сделаем предположение, что '9' подразумевает под собой '2'. Пропуски в данных отсутствуют.

* ***surgery*** - указывает, была ли проведена операция для лечения: 1 = да, была операция. 2 = лечение без операции. Категориальный признак, пропуски составляют 0.3%. Операция для лечения проведена в 60% наблюдаемых случаев.

In [35]:
horse_stat['surgery'].value_counts(normalize = True)

1.0    0.602007
2.0    0.397993
Name: surgery, dtype: float64

* ***rectal_temperature*** - температура лошади, числовой показатель. Согласно пояснительным данным, нормальная температура - 37.8. Размах изучаемых значений составил 5,4 градуса, что выглядит вполне достоверно. Пропуски составляют 20% значений.

* ***pulse*** - пульс лошади, числовой показатель. Согласно пояснительным данным, нормальный пульс взрослой лошади - 30-40 ударов в минуту. Размах данных очень велик, есть экстремальные (подозрительные, возможно ошибочные) выбросы по максимальным значениям(184 уд. в мин), но сторонние источники данных подтверждают, что такие показатели возможны при критических состояниях здоровья. 8% данных отсутствует.

* ***mucous_membranes*** - цвет слизистых оболочек, категориальный признак. Показатели 'normal pink' и 'pale pink' можно расценивать как варианты нормы, то есть 46% лошадей имеют отклонения от нормы по данному показателю. 15,7% данных отсутствует.

In [44]:
horse_stat.mucous_membranes.value_counts(normalize=True)

normal pink              0.312253
pale pink                0.229249
pale cyanotic            0.162055
bright pink              0.118577
bright red / injected    0.098814
dark cyanotic            0.079051
Name: mucous_membranes, dtype: float64

* ***abdominal_distension*** - важный показатель здоровья лошади, содержит информацию о наличии и степени вздутия живота, категориальный признак. Почти 70% лошадей, проходящих лечение, имели в той или иной степени вздутие живота. 
18,7% данных отсутствует.

In [45]:
horse_stat.abdominal_distension.value_counts(normalize=True)

none        0.311475
moderate    0.266393
slight      0.266393
severe      0.155738
Name: abdominal_distension, dtype: float64

* ***packed_cell_volume*** - показатель уровня эритроцитов в крови, числовой признак. Нормальный показатель от 30 до 50. В целом у большинства лошадей показатели в пределах нормы. 9,7% данных отсутствует.

* ***total_protein*** - показатель уровня общего белка, числовой признак. Нормальный показатель от 6 до 7,5 грамм на децилитр. В данных присутствует высокое максимальное значение (89) - сложно понять, допустимо ли такое значение для живой лошади. Возможно, при внесении данных показания могли заносить в других единицах измерения (грамм на литр). Для оценки признака пригодилось бы пояснение поставщика данных или эксперта по лошадям. 11% данных отсутствует.

* ***outcome*** - собственно итог лечения. По результатам лечения выжило 60% пациентов. 0,3% данных отсутствует.

In [46]:
horse_stat.outcome.value_counts(normalize=True)

lived             0.595318
died              0.257525
was euthanized    0.147157
Name: outcome, dtype: float64

***Задание 2. Работа с выбросами***

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

***Порассуждаем о возможных выбросах*** 

In [61]:
for stats in horse_stat_m:
    q1 = horse_stat_m[stats].quantile(0.25)
    q3 = horse_stat_m[stats].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)
    outliers = horse_stat_m.loc[(horse_stat_m[stats] < lower_bound)|
                                       (horse_stat_m[stats] > upper_bound), stats]
    print(stats, outliers, f'Количество выбросов {outliers.count()}', sep='\n')
    print('*' * 45)
    

rectal_temperature
20     39.9
44     35.4
54     40.3
75     39.7
80     36.4
91     40.3
99     39.6
118    36.5
141    36.0
238    36.1
251    36.6
259    40.8
281    40.0
298    36.5
Name: rectal_temperature, dtype: float64
Количество выбросов 14
*********************************************
pulse
3      164.0
41     150.0
55     160.0
255    184.0
275    150.0
Name: pulse, dtype: float64
Количество выбросов 5
*********************************************
packed_cell_volume
4      74.0
144    75.0
151    75.0
Name: packed_cell_volume, dtype: float64
Количество выбросов 3
*********************************************
total_protein
Series([], Name: total_protein, dtype: float64)
Количество выбросов 0
*********************************************


* ***rectal_temperature*** - выбросы не стоит исключать из исследования, так как и минимальная (35.4), и максимальная (40.8) температура может наблюдаться у нездоровой лошади. Вероятнее всего, что в данных ошибки нет.

* ***pulse*** - несмотря на высокие значения показателя, данные выбросы также не стоит исключать из исследования, так как дополнительно просмотренные источники подтверждают, что при неблагополучном течении болезни, пульс может принимать и такие значения (о чем также свидетельствует неблагополучный исход лечения почти всех лошадей с таким пульсом, кроме подозрительно здоровой лошади "5288249")

In [18]:
horse_stat[horse_stat['pulse'] > 140]

case_number    age surgery  rectal_temperature  pulse  \
3        5290409  Young     Yes                39.1  164.0   
39       5277409  Young     Yes                39.2  146.0   
41       5288249  Young      No                39.0  150.0   
55       5282839  Young     Yes                38.6  160.0   
255      5294539  Young     Yes                38.8  184.0   
275      5297159  Young     Yes                38.8  150.0   

          mucous_membranes abdominal_distension  packed_cell_volume  \
3            dark cyanotic               severe                48.0   
39                     NaN                  NaN                 NaN   
41                     NaN                  NaN                47.0   
55   bright red / injected               severe                38.0   
255            normal pink             moderate                33.0   
275          dark cyanotic               slight                50.0   

     total_protein outcome  
3              7.2    died  
39             NaN    died  
41             8.5   lived  
55             NaN    died  
255            3.3    died  
275            6.2    died

* ***packed_cell_volume*** - также не исключаем, так как значения выбросов могут встречаться у нездорового животного.

In [19]:
horse_stat[horse_stat['packed_cell_volume'] > 70]

case_number    age surgery  rectal_temperature  pulse  \
4         530255  Adult      No                37.3  104.0   
35        528812  Adult      No                 NaN  104.0   
49        529642  Adult      No                37.2   84.0   
144       534519  Adult     Yes                37.1   84.0   
151       529399  Adult     Yes                39.3    NaN   
217      5279441  Adult     Yes                38.0   76.0   
231       530028  Adult      No                39.5   92.0   

          mucous_membranes abdominal_distension  packed_cell_volume  \
4            dark cyanotic                  NaN                74.0   
35               pale pink             moderate                73.0   
49   bright red / injected               slight                73.0   
144          dark cyanotic               severe                75.0   
151          dark cyanotic               slight                75.0   
217                    NaN                  NaN                71.0   
231          dark cyanotic                 none                72.0   

     total_protein         outcome  
4              7.4            died  
35             8.4  was euthanized  
49             5.5            died  
144           81.0  was euthanized  
151            NaN            died  
217           11.0           lived  
231            6.4            died

***Как итог*** - значения выбросов лучше не исключать, так как сам датасет предоставляет нам данные о ***больных*** животных, из чего следует, что некоторые из этих данных могут быть далеки от нормы (но в то же время здесь нет абсолютно некорректных данных, которые бы однозначно говорили об их ошибочности).

***Задание 3. Работа с пропусками***

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

В каких данных были пропуски - 
* surgery - 0.3%
* rectal_temperature - 20.0%
* pulse - 8.0%
* mucous_membranes - 15.7%
* abdominal_distension - 18.7%
* packed_cell_volume - 9.7%
* total_protein - 11.0%
* outcome - 0.3%

Чем заменим -

* surgery - мало пропусков, используем моду
* rectal_temperature - много пропусков, смоделируем нормальное распределение вокруг среднего (значения среднего арифметического, моды и медианы практически равны)
* pulse - высокий показатель стандартного отклонения, чтобы избежать искажения данных, применим заполнение средними значениями 
* mucous_membranes - добавим показатель N/A, так как данные категориальные.
* abdominal_distension - добавим показатель N/A, так как данные категориальные.
* packed_cell_volume - близкое значение медианы и среднего арифметического, чтобы избежать искажения данных, применим заполнение средними значениями
* total_protein - чтобы избежать искажения данных, применим заполнение медианными значениями
* outcome - мало пропусков, используем моду

In [20]:
horse_stat['surgery'].fillna(horse_stat['surgery'].mode()[0], inplace = True)

In [21]:
random_values = {}
for index in horse_stat[horse_stat['rectal_temperature'].isnull()].index:
    random_values[index] = np.random.normal(horse_stat['rectal_temperature'].mean(), horse_stat['rectal_temperature'].std())
horse_stat['rectal_temperature'].fillna(random_values, inplace = True)

In [22]:
horse_stat['pulse'].fillna(horse_stat['pulse'].mean(), inplace = True)

In [24]:
horse_stat['mucous_membranes'].fillna('N/A', inplace = True)

In [25]:
horse_stat['abdominal_distension'].fillna('N/A', inplace = True)

In [26]:
horse_stat['packed_cell_volume'].fillna(horse_stat['packed_cell_volume'].mean(), inplace = True)

In [28]:
horse_stat['total_protein'].fillna(horse_stat['total_protein'].median(), inplace = True)

In [27]:
horse_stat['outcome'].fillna(horse_stat['outcome'].mode()[0], inplace = True)

Повторно проверим на пропуски

In [29]:
for column in horse_stat:
    missings = horse_stat[column].isnull().mean()
    print(f'{column} - {missings :.1%}')

case_number - 0.0%
age - 0.0%
surgery - 0.0%
rectal_temperature - 0.0%
pulse - 0.0%
mucous_membranes - 0.0%
abdominal_distension - 0.0%
packed_cell_volume - 0.0%
total_protein - 0.0%
outcome - 0.0%
